libs/ledger-live-common/src/hw/installLanguage.ts

```diff
@@ -0,0 +1,172 @@
+              const { apdu_install_url } = pack;
+              const url = apdu_install_url;
+
+              const { data: rawApdus } = await network({
+                method: "GET",
+                url,
+              });
+
+              const apdus = rawApdus.split(/\r?\n/).filter(Boolean);
+
+              await uninstallAllLanguages(transport);
+
+              for (let i = 0; i < apdus.length; i++) {
+                if (apdus[i].startsWith("e030")) {
+                  subscriber.next({
+                    type: "devicePermissionRequested",
+                  });
+                }
+
+                const response = await transport.exchange(
+                  Buffer.from(apdus[i], "hex")
+                );
+                const status = response.readUInt16BE(response.length - 2);
+                const statusStr = status.toString(16);
+
+                // Some error handling
+                if (status === 0x5501) {
+                  return subscriber.error(
+                    new LanguageInstallRefusedOnDevice(statusStr)
+                  );
+                } else if (status !== 0x9000) {
+                  return subscriber.error(
+                    new TransportError("Unexpected device response", statusStr)
+                  );
+                }
+
+                subscriber.next({
+                  type: "progress",
+                  progress: (i + 1) / apdus.length,
+                });
+              }
+
+              subscriber.next({
+                type: "languageInstalled",
+              });
+
+              subscriber.complete();
+            }),
+            catchError((e: Error) => {
+              if (
+                e instanceof DeviceOnDashboardExpected ||
+                (e &&
+                  e instanceof TransportStatusError &&
+                  [0x6e00, 0x6d00, 0x6e01, 0x6d01, 0x6d02].includes(
+                    // @ts-expect-error typescript not checking agains the instanceof
+                    e.statusCode
+                  ))
+              ) {
+                return from(getAppAndVersion(transport)).pipe(
+                  concatMap((appAndVersion) => {
+                    return !isDashboardName(appAndVersion.name)
+                      ? attemptToQuitApp(transport, appAndVersion)
+                      : of<InstallLanguageEvent>({
+                          type: "appDetected",
+                        });
+                  })
+                );
+              }
+              return throwError(e);
+            })
+          )
+          .subscribe(subscriber);
+
+        return () => {
+          timeoutSub.unsubscribe();
+          sub.unsubscribe();
+        };
+      })
+  );
+
+  return sub as Observable<InstallLanguageEvent>;
+}
+
+const uninstallAllLanguages = async (transport: Transport) => {
+  await transport.send(
+    0xe0,
+    0x33,
+    0xff,
+    0x00,
+    undefined,
+    [0x9000, 0x5501] // Expected responses when uninstalling.
+  );
+};
```

Comment:
Why not move this to another file, perhaps we'll want to uninstall only as part of another flow someday?
Easier to document too if we can link to a single file (personally)

Reply:
I personally like to create the indirection only when it's needed. If we ever use it somewhere else, we can move it.

Reply:
Sure, add the apdu to the doc too 🙏🏼